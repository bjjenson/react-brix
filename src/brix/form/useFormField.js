import { actions } from './fieldReducer'
/**
 * @param  fieldArgs { import("../form").IFormFieldArgs}
 * @returns {import("../form").IFormField}
 */
export const useFormField = (state, dispatch, fieldArgs = {}) => {
  const requiredMessage = fieldArgs.requiredMessage || 'Required'

  const tryValidate = (value, touched) => {
    let result
    if (value && fieldArgs.validate) {
      result = fieldArgs.validate(value, fieldArgs.name, state.get('getAllValues'))
    }

    if (!Boolean(result) && !fieldArgs.optional && touched && value === '') {
      result = requiredMessage
    }

    if (touched) {
      dispatch(actions.validationResult(fieldArgs.name, Boolean(result), result))
    }
    return !Boolean(result)
  }

  const setValue = v => {
    dispatch(actions.updateValue(fieldArgs.name, v))
    tryValidate(v, state.getIn(['current', 'touched']))
  }

  const onChange = ({ target }) => {
    const value = fieldArgs.valueFromTarget ? fieldArgs.valueFromTarget(target) : target.value
    const coercedValue = fieldArgs.normalize ? fieldArgs.normalize(value) : value
    setValue(coercedValue)
  }

  const onBlur = () => {
    dispatch(actions.touched(fieldArgs.name))
    tryValidate(state.getIn(['current', 'value']), true)
  }

  const setValidationResult = result => {
    dispatch(actions.validationResult(fieldArgs.name, true, result))
  }

  const validate = () => {
    return tryValidate(state.getIn(['current', 'value']), true)
  }

  return {
    props: {
      error: state.getIn(['current', 'error']),
      helperText: state.getIn(['current', 'helperText']),
      label: state.getIn(['initial', 'label']),
      value: state.getIn(['current', 'value']),
      // handlers
      onBlur,
      onChange,
    },
    setValidationResult,
    setValue,
    validate,
  }
}
