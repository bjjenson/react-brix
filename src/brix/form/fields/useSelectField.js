import { useFormField } from '../useFormField'

export const useSelectField = (args = {}, initial) => {
  const initialValue = initial || args.value || ''

  const fieldProps = useFormField(initialValue, args)
  fieldProps.props.options = args.options
  return fieldProps
}
