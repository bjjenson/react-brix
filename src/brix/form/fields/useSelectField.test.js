import { useFormField } from '../useFormField'
import { useSelectField } from './useSelectField'

jest.mock('../useFormField')

let initialArgs
beforeEach(() => {
  initialArgs = {
    options: [{ value: 'value1', label: 'label1' }],
  }

  useFormField.mockReturnValue({
    props: {
      prop1: 'prop1',
      prop2: 'prop2',
    },
    validate: 'validateFunction',
  })
})

test('returns all props needed', () => {
  const actual = useSelectField('state', 'dispatch', initialArgs)
  expect(actual).toMatchSnapshot()
  expect(useFormField.mock.calls[0]).toMatchSnapshot()
})
