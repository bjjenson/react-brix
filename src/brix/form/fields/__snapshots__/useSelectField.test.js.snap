// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns all props needed 1`] = `
Object {
  "props": Object {
    "options": Array [
      Object {
        "label": "label1",
        "value": "value1",
      },
    ],
    "prop1": "prop1",
    "prop2": "prop2",
  },
  "validate": "validateFunction",
}
`;

exports[`returns all props needed 2`] = `
Array [
  "state",
  "dispatch",
  Object {
    "options": Array [
      Object {
        "label": "label1",
        "value": "value1",
      },
    ],
  },
]
`;
